sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

```diff
@@ -1103,18 +1127,34 @@ object SQLContext {
         data: Iterator[_],
         beanInfo: BeanInfo,
         attrs: Seq[AttributeReference]): Iterator[InternalRow] = {
-    val extractors =
-      beanInfo.getPropertyDescriptors.filterNot(_.getName == "class").map(_.getReadMethod)
-    val methodsToConverts = extractors.zip(attrs).map { case (e, attr) =>
-      (e, CatalystTypeConverters.createToCatalystConverter(attr.dataType))
-    }
+    val converters = getExtractors(beanInfo, attrs)
     data.map{ element =>
       new GenericInternalRow(
-        methodsToConverts.map { case (e, convert) => convert(e.invoke(element)) }
+        converters.map { case (e, convert) => convert(e.invoke(element)) }
       ): InternalRow
     }
   }
 
+  def getExtractors( beanInfo: BeanInfo,
+                     attrs: Seq[AttributeReference]): Array[(Method, Any => Any)] = {
+   val methodsToConverts = beanInfo.getPropertyDescriptors.filterNot(_.getName == "class").
+      map(_.getReadMethod).zip(attrs)
+    methodsToConverts.map { case (e, attr) =>
+      attr.dataType match {
+        case strct: StructType => {
+          val extractors = getExtractors(Introspector.getBeanInfo(e.getReturnType),
```

Comment:
Looks like it will be better to get the return type from PropertyDescriptor.getPropertyType rather than method return type. The Oracle JDK implementation in java.beans.FeatureDescriptor has this:

```java
    static Class<?> getReturnType(Class<?> base, Method method) {
        if (base == null) {
            base = method.getDeclaringClass();
        }
        return TypeResolver.erase(TypeResolver.resolveInClass(base, method.getGenericReturnType()));
    }
```

Which seems to be doing more than just getting the returnType. So perhaps better option is something like:

```java
beanInfo.getPropertyDescriptors.filterNot(_.getName == "class").zip(attrs) match { case (desc, attr) =>
  attr.dataType match {
    case struct: StructType =>
      val extractors = getExtractors(Introspector.getBeanInfo(desc.getPropertyType),
      ...
    case _ => (desc.getReadMethod, CatalystTypeConverters.createToCatalystConverter(attr.dataType))
  }
}
```